### DISCUSSION TITLE
Extracting and using type of models with computed fields from an extended client.

### DISCUSSION THREAD

**Discussion by johnveeuminga** [prisma/prisma#19947] [Q&A]
Hello, everyone!

Is there a way for us to get and use a type of a model with computed field of an extended client?

For example, I extend the client using this:
```typescript
export const prismaAuction = prisma.$extends({
  result: {
    auction: {
      started: {
        needs: { start_date: true },
        compute(auction) {
          return DateTime.fromJSDate(auction.start_date) <= DateTime.now()
        }
      },
      ended: {
        needs: { end_date: true },
        compute(auction) {
          console.log(DateTime.fromJSDate(auction.end_date))
          return DateTime.fromJSDate(auction.end_date) <= DateTime.now()
        }
      },
      ongoing: {
        needs: { end_date: true, start_date: true},
        compute(auction) {
          return DateTime.fromJSDate(auction.end_date) > DateTime.now()
            && DateTime.fromJSDate(auction.start_date) <= DateTime.now()
        }
      }
    },
  }
});
```

Is there anyway we can get the generated model types without manually defining the type? Something like in the docs? See below for reference:

```typescript
import { Prisma } from '@prisma/client'

// 1: Define a type that includes the relation to `Post`
const userWithPosts = Prisma.validator<Prisma.UserArgs>()({
  include: { posts: true },
})

// 2: Define a type that only contains a subset of the scalar fields
const userPersonalData = Prisma.validator<Prisma.UserArgs>()({
  select: { email: true, name: true },
})

// 3: This type will include a user and all their posts
type UserWithPosts = Prisma.UserGetPayload<typeof userWithPosts>
```

**Comment by johnveeuminga** [comment]
For context, I get the record like this.
```typescript
export async function getAuctions(params: Prisma.AuctionFindManyArgs = {}) { 
  const auctions = await prismaAuction.auction.findMany(params);

  return auctions;
}
```

Now on a React component, I loop through the results and pass a function that calculates classes to be used on the badge. 
```typescript
     const auctions = await getAuctions()
     //  additional code here
     {
          auctions && auctions.map(auction => (
            <tr key={auction.id}>
              <td>
                <span className={`badge ${performClassCalculation(auction)}`}>
                 { performNameCalculation(auction) }
                </span>
              </td>
            </tr>
          ))
    }
```

The `performClassCalculation` and `performNameCalculation` are functions that needs the `auction` as a param like this:
```typescript
 function extractStatusClassName(auction: any) { // I need the generated type here
   // Perform calculation here  
 }
```

**Reply by millsp** [reply]
Does this comment help you?
https://github.com/prisma/prisma/issues/19112#issuecomment-1605246042

**Reply by nurul3101** [reply]
Hey @johnveeuminga 
Did you had chance to look at Pierre's suggestion?

Please let us know in case you still need any assistance. If not, please consider marking the discussion as Answered/Closed.

**Comment by lugrinder** [comment]
What will be the equivalent of a `Select` with an extension?

For example, for a `Model` with `ModelSelect` generated by Prisma:

```typescript
export type ModelSelect<ExtArgs extends $Extensions.Args = $Extensions.DefaultArgs> = $Extensions.GetSelect<{
  id?: boolean
  name?: boolean
  _count?: boolean | ModelCountOutputTypeArgs<ExtArgs>
}, ExtArgs["result"]["device"]>
```

I can create a "validated" `select` with:

```typescript
const select = Prisma.validator<Prisma.ModelSelect>()({
    id: true,
    name: true,
})
```

Which is the equivalent for extended version?

```typescript
export const extendedClient = client.$extends({
    result: {
        model: {
            upper: {
                needs: { name: true },
                compute (model): {
                   return model.name.toUpperCase()
                }
            }
        }
    }
})

const extendedSelect = Prisma.validator<???>()({
    id: true,
    name: true,
    upper: true
})
```

**Reply by millsp** [reply]
Hey @lugrinder you can use the `Prisma.validator` for this, where `prisma` is either your client or an extended client

```ts
const data4 = Prisma.validator(prisma, 'user', 'findFirst', 'select')({
  id: true,
})
```

**Reply by lugrinder** [reply]
Thanks for your suggestion, but I have an issue with this if I use the extended client, with the base client works fine. If I return the result of the validator in a function, for example in a constructor function to build a select based on function arguments like this:

```typescript
function buildSelect(name: boolean) {
  return Prisma.validator(prisma, 'user', 'findFirst', 'select')({
    id: true,
    name,
  })
}

```
I get the next error:

`TS7023: 'buildSelect' implicitly has return type 'any' because it does not have a return type annotation and is referenced directly or indirectly in one of its return expressions.`

And, how can I define a type for the result payload of an extended model? Now I can get this type with:

```typescript
type myPayloadType = Prisma.MyModelGetPayload<typeof mySelect>
```

Using the returned type of `buildSelect()` for `mySelect`.

**Reply by millsp** [reply]
If the following does not work for you, then we would appreciate if you could open a bug report with a reproduction. On my end, this works fine:

```ts
const prisma = new PrismaClient().$extends({
  result: {
    user: {
      test: {
        needs: {
          email: true,
        },
        compute(data) {
          return data.email
        },
      }
    }
  }
})

const a = Prisma.validator(prisma, 'user', 'findFirst', 'select')({
  id: true,
  test: true
})
```

**Reply by millsp** [reply]
With and without extension, you can now get your types with the new `Result` helper

```ts
type test = Prisma.Result<typeof prisma.user, { select: { name: true } }, 'findFirstOrThrow'>
```

**Reply by lugrinder** [reply]
Using the `Result` helper with the extended client, I get the next error:

`TS2456: Type alias 'test' circularly references itself.`

With base client works fine, only when using extended client...

I'll try to create a fully running example that covers both errors. Any recommended codepen template, or similar service, for PrismaJS examples?

**Reply by millsp** [reply]
Huh, that is interesting. I think we had a similar report elsewhere but could not get a reproduction from the user. It might be very schema specific, so I'd really love to get your reproduction, thanks for taking the time! A simple github repo will be enough, barebones, I don't mind if it isn't perfect, as I will transform that into a test most likely later on with the fix. Thank you.

**Reply by lugrinder** [reply]
I’ve found the problem. It’s not an issue with Prisma; it’s an error on my part. I’m migrating from my custom transform function that adds computed fields. During the adaption, I made functions for the computed fields that were shared between `$extend` and my transformation function. Those shared functions use the type of returned value by the `select` as an argument, which creates a circular dependency.
Thanks for all...

**Reply by millsp** [reply]
You're welcome, happy you could unblock your situation.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
